Support for the Research Organization Registry (ROR)

joemull opened the pull request:
Closes #3168.

## Overview
This piece of work adds a ROR data model, including affiliations, organizations, organization names, and locations. It includes an import routine to fetch and process ROR's full database. I've included as much backwards compatibility as possible for importers that do not know about ROR. It also adds a user interface for editing one's own affiliations, and an admin interface for all of the relevant models.

## Command line interface

```sh
❯ python src/manage.py import_ror_data
DEBUG 2024-12-09 18:41:33,666 connectionpool P:131157 T:135176669522816 Starting new HTTPS connection (1): zenodo.org:443
DEBUG 2024-12-09 18:41:34,697 connectionpool P:131157 T:135176669522816 https://zenodo.org:443 "GET /api/communities/ror-data/records?sort=newest HTTP/1.1" 200 None
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 101.06it/s]
INFO 2024-12-09 18:41:38,658 import_ror_data P:131157 T:135176669522816 successful
```

## User interface

![image](https://github.com/user-attachments/assets/9a5e98c4-bee3-4e9d-a907-bff6c38f4ada)
![image](https://github.com/user-attachments/assets/71700c9f-6ccf-4544-819a-98a3da9ba653)
![image](https://github.com/user-attachments/assets/16918879-93d9-4e2a-b929-8a36fc0fd9e4)
![image](https://github.com/user-attachments/assets/bd517c8a-3b54-40d6-bd8e-3903c48d7bbb)
![image](https://github.com/user-attachments/assets/62bdee29-ba3e-424e-94b6-e801a6d5842e)
![image](https://github.com/user-attachments/assets/3d2239b1-586e-479b-907a-e6bdc344e183)
![image](https://github.com/user-attachments/assets/08065abe-df16-43b1-aedd-a232b87d9bcc)

MartinPaulEve commented:
If you enter an organization name that isn't found (e.g., in test import, "Birkbeck") and then click on one of the pagination items at the bottom, you are redirected to an erroring page eg:

http://127.0.0.1:8000/JRNL/profile/organization/search/?q=birkbeck&page=3&paginate_by=25

This errors because the item isn't found, but the search query is inserted into the pagination link

MartinPaulEve commented on src/submission/tests.py:
```diff
@@ -66,28 +66,32 @@ def create_journal():
             'salutation': 'Prof.',
             'first_name': 'Martin',
+            'middle_name': '',
             'last_name': 'Eve',
```
Top test author

joemull replied:
Without a doubt!

MartinPaulEve commented on src/utils/migrations/0035_rorimport_rorimporterror.py:
```diff
@@ -0,0 +1,37 @@
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('started', models.DateTimeField(auto_now_add=True)),
+                ('stopped', models.DateTimeField(blank=True, null=True)),
+                ('status', models.CharField(choices=[('ongoing', 'Ongoing'), ('unnecessary', 'Unnecessary'), ('successful', 'Successful'), ('failed', 'Failed')], default='ongoing')),
```
Should these choices reference RORImportStatus rather than duplicating?

joemull replied:
I believe this is the standard way Django records choices in migrations, even when the choices are expressed in a class. The reason is that if the migration referenced a variable in the models file, then if the choices list ever changed there, the migration might stop working.

Example:
https://github.com/openlibhums/janeway/blob/bc8a3279b15d2a88d6390c4136655fcd83056ca4/src/journal/migrations/0031_issue_old_issue_type.py#L35

MartinPaulEve reviewed (changes requested):
This mostly looks really good. Just a few small things.

- [ ] Pagination in ROR search currently redirects to page not found error
- [x] List of affiliations should show which is primary?
- [ ] RORImportStatus should not be duplicated?

joemull commented on src/core/logic.py:
```diff
@@ -868,7 +868,7 @@ def check_for_bad_login_attempts(request):
 
     attempts = models.LoginAttempt.objects.filter(user_agent=user_agent, ip_address=ip_address, timestamp__gte=time)
-    print(time, attempts.count())
+    logger.info(time, attempts.count())
```
Yes probably, though I'm just removing the noisy print statement here, not changing this feature on this branch.

joemull commented:
> * [ ]  Pagination in ROR search currently redirects to page not found error

This turned out to be a Django bug I believe. I think it's obscure because most people don't encounter it, since they submit their search query to change the list of items before selecting a page.

But I've gone ahead and fixed it for this view.

> * [ ]  List of affiliations should show which is primary?

Yes, good call. I added a label.

I also updated the save method to make the affiliation primary if it's the first one.

![image](https://github.com/user-attachments/assets/fa35b8a6-bc67-4c8b-be71-29cc3e791b1a)

> * [ ]  RORImportStatus should not be duplicated?

That's standard, I'm pretty sure. See comment inline.

joemull commented:
Martin approved this off GitHub on 13 Dec. Moving it along to @ajrbyers and @mauromsl.

ajrbyers commented on src/core/models.py:
```diff
@@ -1906,3 +1923,558 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+        try:
+            organization = cls.objects.get(labels__value=value)
+            created = False
+        except cls.DoesNotExist or cls.MultipleObjectsReturned:
```
This wont catch `cls.MultipleObjectsReturned`. It should be a tuple `except (cls.DoesNotExist, cls.MultipleObjectsReturned):`

ajrbyers commented on src/core/models.py:
```diff
@@ -1906,3 +1923,558 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+            try:
+                organization = cls.objects.get(aliases__value=value)
+                created = False
+            except cls.DoesNotExist or cls.MultipleObjectsReturned:
```
Same issue here. This should be a tuple.

ajrbyers commented on src/core/models.py:
```diff
@@ -1906,3 +1923,558 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+                        ror__exact='',
+                    )
+                    created = False
+                except cls.DoesNotExist or cls.MultipleObjectsReturned:
```
Same issue here. This should be a tuple.

ajrbyers commented on src/core/models.py:
```diff
@@ -1906,3 +1923,558 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+        # canonical name data from ROR (e.g. labels)?
+        try:
+            organization = cls.objects.get(labels__value=value)
+            created = False
```
If an exception is raised `created` wont be set. Should it be set before the try block?

ajrbyers commented on src/templates/admin/elements/layout/key_value_above.html:
Nice.

ajrbyers reviewed (changes requested):
This is generally very good. The CBVs are growing on me... slowly.

There are some comments inline that should be addressed. Here are some general comments in addition:

- Institution and Affiliation fields are removed without preservation.
- Mauro made a suggestion to handle setting inst/dept etc using the Manager, this should be added.
- When my affiliation was snapshotted it did not include the start/end date or the title.
- There is a conflict currently listed


<img width="1236" alt="Screenshot 2025-01-15 at 10 52 31" src="https://github.com/user-attachments/assets/d94e9ed5-4c0d-42f6-bb32-88fd0f80402b" />

mauromsl commented on src/utils/management/commands/import_ror_data.py:
```diff
@@ -0,0 +1,53 @@
+                 'Pass --test_full_import to import the entire dump in development.',
+            action='store_true',
+        )
+        return super().add_arguments(parser)
```
It is still good practice to call `super()` when overriding a method like this, in case the underlying implementation changes or a mixin is added in the inheritance chain

mauromsl commented on src/core/views.py:
```diff
@@ -2775,3 +2780,153 @@ def post(self, request, *args, **kwargs):
+    Allows a user to update one of their own affiliations.
+    """
+
+    model = core_models.Affiliation
+    pk_url_kwarg = 'affiliation_id'
+    form_class = forms.AffiliationForm
+    success_url = reverse_lazy('core_edit_profile')
+    success_message = _("Affiliation updated: %(organization)s")
+
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+        context['account'] = self.request.user
+        affiliation_id = self.kwargs.get('affiliation_id')
+        context['affiliation'] = core_models.Affiliation.objects.get(
+            pk=affiliation_id
+        )
+        context['organization'] = context['affiliation'].organization
+        return context
+
+
+@method_decorator(login_required, name='dispatch')
+class AffiliationDeleteView(SuccessMessageMixin, DeleteView):
+    """
+    Allows a user to delete one of their own affiliations.
+    """
+
+    model = core_models.Affiliation
+    pk_url_kwarg = 'affiliation_id'
+    success_url = reverse_lazy('core_edit_profile')
+    success_message = _("Affiliation deleted")
+
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+        context['account'] = self.request.user
+        affiliation_id = self.kwargs.get('affiliation_id')
+        context['affiliation'] = core_models.Affiliation.objects.get(
+            pk=affiliation_id
+        )
+        context['organization'] = context['affiliation'].organization
+        return context
```
Have we discussed introducing the use of Django's generic CBVs?
They don't seem to provide any value in terms of reusability, have added cognitive load over existing views and more importantly, they are inconsistent with the rest of the codebase

joemull replied:
> Have we discussed introducing the use of Django's generic CBVs? They don't seem to provide any value in terms of reusability, have added cognitive load over existing views and more importantly, they are inconsistent with the rest of the codebase

Not sure I agree that "they don't seem to provide any value in terms of reusability". Let's schedule a time to talk about your concerns.

ajrbyers commented:
@joemull running the migrations on a clean OLH database gives me:

```python
Running migrations:
  Applying utils.0035_rorimport_rorimporterror... OK
  Applying core.0100_location_organization_affiliation... OK
  Applying core.0101_migrate_affiliation_institution...Traceback (most recent call last):
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.StringDataRightTruncation: value too long for type character varying(200)


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ajrbyers/Code/janeway/src/manage.py", line 16, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 356, in handle
    post_migrate_state = executor.migrate(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/Users/ajrbyers/Code/janeway/src/core/migrations/0101_migrate_affiliation_institution.py", line 71, in migrate_affiliation_institution
    create_affiliation(
  File "/Users/ajrbyers/Code/janeway/src/core/migrations/0101_migrate_affiliation_institution.py", line 47, in create_affiliation
    organization = create_organization(apps, old_institution, old_country)
  File "/Users/ajrbyers/Code/janeway/src/core/migrations/0101_migrate_affiliation_institution.py", line 30, in create_organization
    OrganizationName.objects.create(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/query.py", line 658, in create
    obj.save(force_insert=True, using=self.db)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/base.py", line 814, in save
    self.save_base(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/base.py", line 877, in save_base
    updated = self._save_table(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/base.py", line 1020, in _save_table
    results = self._do_insert(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/base.py", line 1061, in _do_insert
    return manager._insert(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/query.py", line 1805, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1822, in execute_sql
    cursor.execute(sql, params)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/backends/utils.py", line 102, in execute
    return super().execute(sql, params)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/Users/ajrbyers/.venvs/janeway/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.DataError: value too long for type character varying(200)
```

Looks like `OrganizationName.value` isn't large enough for the old institution field:

https://github.com/openlibhums/janeway/blob/d9c1ea56fcb2a0e12c5a8798e0518320e36e3da7/src/core/models.py#L277-L282

ajrbyers reviewed (changes requested):
See comment on migration error.

joemull commented:
OK expanded it @ajrbyers

ajrbyers commented on src/review/logic.py:
```diff
@@ -824,24 +824,22 @@ def process_reviewer_csv(path, request, article, form):
         reader = csv.DictReader(csv_file)
         reviewers = []
         for row in reader:
-            try:
-                country = core_models.Country.objects.get(code=row.get('country'))
-            except core_models.Country.DoesNotExist:
-                country = None
-
```
This is removed and not handled so this data is lost.

joemull replied:
Thanks for catching. Again an issue with the old backwards compatibility approach.

ajrbyers commented on src/utils/models.py:
```diff
@@ -3,10 +3,12 @@
 __license__ = "AGPL v3"
 __maintainer__ = "Birkbeck Centre for Technology and Publishing"
 
-import json as jason
```
Change this back......................... JASON! 😄

joemull replied:
Who is Jason and what did he do to Janeway???? 😄

ajrbyers reviewed (changes requested):
Thanks for this, Joe. Its looking good. I've found a couple of bugs that need to be squashed and added a couple of comments for you to look over.

joemull commented on src/utils/testing/helpers.py:
```diff
@@ -62,6 +62,7 @@ def create_user(username, roles=None, journal=None, **attrs):
             journal=journal
         )
 
+    user.save()
```
Thanks for catching that. It was an earlier attempt at backward compatibility before writing the ModelManager.

joemull commented on src/core/forms/forms.py:
```diff
@@ -940,3 +940,33 @@ class AccountRoleForm(forms.ModelForm):
+        fields = ('value',)
+
+
+class AffiliationForm(forms.ModelForm):
+
+    class Meta:
+        model = models.Affiliation
+        fields = '__all__'
+        widgets = {
+            'account': forms.HiddenInput,
+            'frozen_author': forms.HiddenInput,
+            'preprint_author': forms.HiddenInput,
+            'organization': forms.HiddenInput,
+            'start': HTMLDateInput,
+            'end': HTMLDateInput,
+        }
```
Ah, see what you're saying now. I'll keep working on it and remove the account-related fields from the form. I'll use a different form for the frozen author editing interface, but that one probably also won't include account.

joemull replied:
Blocked by https://github.com/openlibhums/janeway/issues/4625

joemull replied:
Unblocked and changes made.

ajrbyers commented on src/core/forms/forms.py:
```diff
@@ -954,20 +954,33 @@ class Meta:
+    def save(self, commit=True):
+        affiliation = super().save(commit=False)
+        request = get_current_request()
+        affiliation.account = request.user
```
This works but why not just pass the user object in like organization?

mauromsl commented on src/utils/models.py:
```diff
@@ -348,3 +354,170 @@ def fetch(url, up_auth_file='', up_base_url='', ojs_auth_file=''):
+        for record in tqdm.tqdm(ror_data, desc="Finding new or updated ROR records"):
+            ror = record.get("id", {})
+            last_modified = record.get("admin", {}).get("last_modified", {})
+            timestamp = last_modified.get("date", "")
+            if ror and timestamp and timestamp > timestamps.get(ror, ''):
```
Can we make sure the defaults for these `.get()` are also floats for explicitness? we can go with `timestamps.get(ror, float())` or `timestamps.get(ror, float('-inf'))` if the timestamp could somehow be 0.0

joemull replied:
This timestamp will always be a string, actually. It's just `YYYY-MM-DD` from the ROR JSON, no Python object being created. Is that alright?

mauromsl reviewed (changes requested):
Thanks Joe, lots of work went into this one I can tell.

some important comments regarding backwards compatibility, there rest are just some polish and nitpicking,but very happy with the work at large here

joemull commented on src/submission/models.py:
```diff
@@ -1930,6 +1930,31 @@ def pinned(self):
 
+class FrozenAuthorManager(models.Manager):
+
+    def get_or_create(self, defaults=None, **kwargs):
+        """
+        Backwards-compatible override for affiliation-related kwargs
+        """
+        # check for deprecated fields related to affiliation
+        institution = kwargs.pop('institution', '')
+        department = kwargs.pop('department', '')
+        country = kwargs.pop('country', '')
+
+        frozen_author, created = super().get_or_create(defaults, **kwargs)
+
+        # create or update affiliation
+        if institution or department or country:
+            Affiliation.naive_get_or_create(
+                institution=institution,
+                department=department,
+                country=country,
+                frozen_author=frozen_author,
+            )
+
+        return frozen_author, created
```
@mauromsl I can expand this, but I believe it will also need to include get methods, because if you pass a field to `get_or_create` that does not exist on the model, the lookup will fail: `django.core.exceptions.FieldError: Cannot resolve keyword 'country' into field.` OK with you if I also override `get` for all three models in addition to overriding `create` and `update`?

joemull commented on src/cron/management/commands/send_publication_notifications.py:
```diff
@@ -47,12 +50,12 @@ def handle(self, *args, **options):
                 journal=journal,
             ).value:
-                print('Sending notification for {}'.format(journal.name))
+                logger.info('Sending notification for {}'.format(journal.name))
```
I believe I changed this so that they would not interrupt tests. Is there something wrong with logger.info here?

joemull commented on src/core/views.py:
```diff
@@ -2800,3 +2805,221 @@ def post(self, request, *args, **kwargs):
+    organization = core_models.Organization.objects.get(
+        pk=organization_id,
+    )
+    form = forms.AccountAffiliationForm(
+        account=request.user,
+        organization=organization,
+    )
+
+    if request.method == 'POST':
+        form = forms.AccountAffiliationForm(
+            request.POST,
+            account=request.user,
+            organization=organization,
+        )
+        if form.is_valid():
+            affiliation = form.save()
+            messages.add_message(
+                request,
+                messages.SUCCESS,
+                _("Affiliation created: %s") % affiliation,
+            )
+            return redirect(reverse('core_edit_profile'))
```
Sure, I've added a clean method.

joemull commented:
OK @mauromsl, I've addressed some of the smaller things but have questions about others.

joemull commented:
Two more things in addition to the above comments, @mauromsl:

- I've updated the exclusive fields check to use the constraint like I did on the credit branch, but also added an option for all fields to be blank, which is a logical conclusion of their individual `blank=True` settings. I'll update the credit branch as well.
- I've gone ahead and moved storage of ROR IDs to the short, non-URL form, even though you didn't mention it in your review, because of our conversation the other day about storing URLs in databases. This has a slight UX hit in that you can't search by the full ROR URL. But that is not a big use case, and we don't really want end-users dealing with raw ROR IDs anyway, so I can live with it.

mauromsl commented on src/core/models.py:
```diff
@@ -230,6 +236,28 @@ def create_superuser(self, email, password, **kwargs):
+        Backwards-compatible override for affiliation-related kwargs
+        """
+        # check for deprecated fields related to affiliation
+        institution = kwargs.pop('institution', '')
+        department = kwargs.pop('department', '')
+        country = kwargs.pop('country', '')
```
we need to check and pop these keys from the `defaults` argument as well.

mauromsl commented on src/core/models.py:
```diff
@@ -230,6 +236,28 @@ def create_superuser(self, email, password, **kwargs):
     def get_queryset(self):
         return AccountQuerySet(self.model)
 
+    def get_or_create(self, defaults=None, **kwargs):
+        """
+        Backwards-compatible override for affiliation-related kwargs
+        """
```
we should instead do this at the Queryset level, since django will delegate this call to `AccountQuerySet` and the queryset also exposes these ORM operations.

Also, don't forget about `get`, `create`, `filter` and `update_or_create`

The bulk operations can be ignored IMO

mauromsl commented on src/core/models.py:
```diff
@@ -1919,3 +1968,641 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+        null=True,
+        help_text="This name is a ROR-provided acronym.",
+    )
+    language = models.CharField(
+        max_length=10,
+        blank=True,
+        choices=submission_models.LANGUAGE_CHOICES,
+    )
```
Is there a rationale behind avoiding the use of `django-modeltranslations` here? Does it have to do with the need for a distinction between translations and `ror_display_for`?

joemull replied:
Thanks for raising this. Yes, ROR data sometimes has multiple names per language. I've written it up here: https://github.com/openlibhums/janeway/issues/4657

Coincidentally, the importer was bugged because ROR gives 2-character language codes but our `LANGUAGE_CHOICES` uses 3-character codes. So I've fixed that by using a small ISO-639 utility. I've also documented which version of ISO-639 we hardcode, for future reference.

mauromsl commented on src/core/models.py:
```diff
@@ -1919,3 +1968,641 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+
+
+def validate_ror_id(ror_id):
+    ror_regex = '^0[a-hj-km-np-tv-z|0-9]{6}[0-9]{2}$'
```
It is good practice to [compile the regex](https://docs.python.org/3/library/re.html#re.compile) explicitly as a constant, both for performance and reusability. Also, remember to flag the string as a regex with `r'^0[a-hj-km-np-tv-z|0-9]{6}[0-9]{2}$'`

mauromsl commented on src/core/models.py:
```diff
@@ -1919,3 +1968,641 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+    @classmethod
+    def get_primary(
+        cls,
+        account=None,
+        frozen_author=None,
+        preprint_author=None,
```
I don't think we ever want three arguments and instead we can use a single mixed-typed argument (e.g. 'author') which can be an instance of a Model that has the `affiliation_set` interface

mauromsl commented on src/core/models.py:
```diff
@@ -1919,3 +1968,641 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+            affiliation.department = value
+            affiliation.save()
+        except ValueError:
+            logger.warn('The department could not be set.')
```
if we use a warning level, we need more context on this message (e.g.   `logger.warn('The department could not be set to "%s"' % value)`

mauromsl commented on src/core/models.py:
```diff
@@ -1919,3 +1968,641 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+            )
+            affiliation.department = value
+            affiliation.save()
+        except ValueError:
```
Given how we are running the query, we also now need to handle `django.db.IntegrityError` for the `check_exclusive_fields_constraint` declaration on `Affiliation`

mauromsl commented on src/core/models.py:
```diff
@@ -1919,3 +1968,641 @@ def log_hijack_ended(sender, hijacker, hijacked, request, **kwargs):
+        null=True,
+        on_delete=models.SET_NULL,
+    )
+    latitude = models.DecimalField(
+        max_digits=9,
+        decimal_places=6,
+        blank=True,
+        null=True,
+    )
+    longitude = models.DecimalField(
+        max_digits=9,
+        decimal_places=6,
+        blank=True,
+        null=True,
+    )
```
Are we making use of these fields? If not, I reckon it is worth holding off on exposing an API that we might need to later make backwards compatible if we end up using [PostGIS](https://postgis.net/), similar to the ongoing [KOMET](https://github.com/openlibhums/janeway/issues/1928) project

joemull replied:
Yeah, good point. Removed these fields.

mauromsl commented on src/cron/management/commands/send_publication_notifications.py:
```diff
@@ -47,12 +50,12 @@ def handle(self, *args, **options):
 
                 if bcc_list:
-                    print("Sending notifications to {}".format(
+                    logger.info("Sending notifications to {}".format(
```
we still have a few messages not being received by `stdout`

mauromsl commented on src/submission/models.py:
```diff
@@ -1932,6 +1932,31 @@ def pinned(self):
 
+class FrozenAuthorManager(models.Manager):
+
+    def get_or_create(self, defaults=None, **kwargs):
+        """
+        Backwards-compatible override for affiliation-related kwargs
+        """
+        # check for deprecated fields related to affiliation
+        institution = kwargs.pop('institution', '')
+        department = kwargs.pop('department', '')
+        country = kwargs.pop('country', '')
+
+        frozen_author, created = super().get_or_create(defaults, **kwargs)
+
+        # create or update affiliation
+        if institution or department or country:
+            Affiliation.get_or_create_without_ror(
+                institution=institution,
+                department=department,
+                country=country,
+                frozen_author=frozen_author,
+            )
+
+        return frozen_author, created
+
```
as with `Account`, we should be thorough here and make sure we cover the main ORM methods and do so at the `Queryset` level

mauromsl commented on src/submission/tests.py:
```diff
@@ -51,8 +51,8 @@ def test_new_journals_has_submission_configuration(self):
 
-    @staticmethod
-    def create_journal():
+    @classmethod
+    def create_journal(cls):
```
what does this change achieve? the `cls` argument is unused

mauromsl reviewed (changes requested):
Mega job here @joemull 

I've added a few comments regarding backwards compatibility and database integrity but very happy overall.

I'm testing out the ROR data import procedure locally to try and anticipate any possible hiccups in production but I don't want to hold you up further, good work!

mauromsl reviewed (changes requested):
Awesome! a couple minor nitpicks and we are good to go. This is turning into a very nice piece of kit :)
